crates/aggchain-proof-core/src/proof.rs

```diff
@@ -21,24 +24,49 @@ pub struct AggchainProofWitness {
     pub prev_local_exit_root: Digest,
     /// New local exit root.
     pub new_local_exit_root: Digest,
-    /// Commitment to the imported bridge exits indexes.
-    pub commit_imported_bridge_exits: Digest,
     /// L1 info root used to import bridge exits.
     pub l1_info_root: Digest,
     /// Origin network for which the proof was generated.
     pub origin_network: u32,
     /// Full execution proof with its metadata.
-    pub fep: FepWithPublicValues,
+    pub fep: FepPublicValues,
+    /// L1 info tree leaf and index containing the `l1Head` as block hash.
+    pub l1_info_tree_leaf: (u32, L1InfoTreeLeaf),
+    /// Inclusion proof of the leaf to the l1 info root.
+    pub l1_head_inclusion_proof: LETMerkleProof<Keccak256Hasher>,
+    /// List of the global index of each imported bridge exit.
+    pub global_indices: Vec<B256>,
     /// Bridge witness related data.
     pub bridge_witness: BridgeWitness,
 }
 
 impl AggchainProofWitness {
     pub fn verify_aggchain_inputs(&self) -> Result<AggchainProofPublicValues, ProofError> {
-        // Verify the FEP exclusively within the SP1 VM
-        #[cfg(target_os = "zkvm")]
+        // Verify the FEP proof or ECDSA signature.
         self.fep.verify()?;
 
+        // Verify that the `l1Head` considered by the FEP exists in the L1 Info Tree
+        {
+            if self.fep.l1_head != self.l1_info_tree_leaf.1.block_hash {
+                return Err(ProofError::MismatchL1Head {
+                    from_l1_info_tree_leaf: self.l1_info_tree_leaf.1.block_hash,
+                    from_fep_public_values: self.fep.l1_head,
+                });
+            }
+
+            if !self.l1_head_inclusion_proof.verify(
+                self.l1_info_tree_leaf.1.hash(),
+                self.l1_info_tree_leaf.0,
+                self.l1_info_root,
+            ) {
```

Comment:
Tiny style nit. I prefer to avoid multi-line if conditions.
```suggestion
            let inclusion_proof_valid = self.l1_head_inclusion_proof.verify(
                self.l1_info_tree_leaf.1.hash(),
                self.l1_info_tree_leaf.0,
                self.l1_info_root,
            );
            if !inclusion_proof_valid {
```